### PR TITLE
os/bluestore: be flexible in choosing allocation size

### PR DESCRIPTION
**Pull request by ifed01**
Hence we can use both small allocations to save space and optimal one to provide speed.
This patch implements both general framework implementation and small allocations application for both objects with expected object size below some threshold (48K for now) and pools configured with optimize_for_size=true mode.

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -6299,6 +6307,20 @@ int BlueStore::mkfs()
       min_alloc_size = cct->_conf->bluestore_min_alloc_size_ssd;
     }
   }
+  // choose optimal_alloc_size
+  if (cct->_conf->bluestore_optimal_alloc_size) {
+    optimal_alloc_size = cct->_conf->bluestore_optimal_alloc_size;
+  }
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc]: nit: elses should be on same line as }

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -13598,6 +13704,29 @@ void BlueStore::_choose_write_options(
...
+  if (!o->onode.use_small_alloc() &&
+    min_alloc_size != optimal_alloc_size && o->onode.size == 0) {
+    // enable small allocations for tiny object which size is known beforehand
+    if (o->onode.expected_object_size &&
+        o->onode.expected_object_size <= (optimal_alloc_size * 3 / 4)) { //FIXME minor: configure threshold?
+      o->onode.set_flag(bluestore_onode_t::FLAG_SMALL_ALLOC);
+      wctx->alloc_size = min_alloc_size;
+      alloc_size_order = min_alloc_size_order;
+    }
+    // more cases to follow
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc]: it looks like RBD is the only client to set the hint right now, and with its object size isn't likely to benefit from this.

Sage mentioned on-list that RGW sends IMMUTABLE flag hints, but I don't see it doing that - that seems like an easy addition. Given this existing check in bluestore, rgw could set these corresponding flags to prefer larger blobs too: https://github.com/ceph/ceph/blob/8b9ffa696dc30750cd64baaac90590e3c5ef643a/src/os/bluestore/BlueStore.cc#L13610-L13614

**Comment by jdurgin** [comment]
Seems like a reasonable structure. @liewegas am I missing where rgw is using the hints?

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: @jdurgin - yeah, for now this PR is mostly about alloc size selection framework not its usage. The followup is to extend it with set_alloc_hint call for rgw/rados and provide expected_object_size whenever possible. E.g. https://github.com/ceph/ceph/pull/30056/files#diff-c24acfa82f15c38ae9370cab8630a579
performs that. Not sure about setting IMMUTABLE flag, IMO it's redundant at this point as I'd like to use expected_object_size as a trigger  instead.

Comment (jdurgin) [reply src/os/bluestore/BlueStore.cc]: expected_object_size makes sense for the alloc size determination alone, I agree. It seems like IMMUTABLE would be useful for other areas, like compression, dedup, perhaps tiering. Not necessary for this PR though.

File: src/rgw/rgw_putobj_processor.cc [hunk 3]
```diff
@@ -88,6 +88,8 @@ int RadosWriter::process(bufferlist&& bl, uint64_t offset)
   }
   librados::ObjectWriteOperation op;
   if (offset == 0) {
+    uint64_t len = data.length();
```
Comment (jdurgin) [inline src/rgw/rgw_putobj_processor.cc]: @cbodley @yehudasa is this commit sufficient, or are there any more places where we know the object size and could hint it?

Comment (cbodley) [reply src/rgw/rgw_putobj_processor.cc]: this is sufficient to cover the immutable tail objects

the mutable head object writes happen in `RGWRados::Object::Write::_do_write_meta()` with `op.write_full(*meta.data);` but it's not clear to me whether we should add the hint there too

Comment (jdurgin) [reply src/rgw/rgw_putobj_processor.cc]: the idea is this would help particularly for small objects - would those only have a head, and no tail objects?

Comment (cbodley) [reply src/rgw/rgw_putobj_processor.cc]: exactly right, small object data would always go in the head only. if a head object gets overwritten, we would recreate the object with a new size hint. that's expected to work, right?

**Comment by ifed01** [comment]
rebased

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (liewegas) [reply src/os/bluestore/BlueStore.cc]: IMMUTABLE is important for choosing the csum block size...

**Comment by ifed01** [comment]
We don't need this any more as bluestore_min_alloc_size_hdd defaults to 4K now, see https://github.com/ceph/ceph/pull/34588

**Pull request closed without merging**